[pythonic resources] Make init context available in Pythonic resources

benpankow opened the pull request:
## Summary

Makes the `InitResourceContext` available to Pythonic resources through a `get_context()` getter. This can be used to retrieve e.g. instance information in the body of a resource.

```python
 class OutputDirResource(ConfigurableResource):
    output_dir: Optional[str] = None

    def get_effective_output_dir(self) -> str:
        if self.output_dir:
            return self.output_dir

        context = self.get_resource_context()
        assert context.instance
        return context.instance.storage_directory()
```

## Test Plan

Added unit test, todo a few more

benpankow commented:
Current dependencies on/for this PR:
* master
  * **PR #13613** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13613" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/13613?utm_source=stack-comment).

schrockn commented on python_modules/dagster/dagster/_config/pythonic_config/__init__.py:
```diff
@@ -706,6 +713,13 @@ def initialize_and_run(self, context: InitResourceContext) -> TResValue:
     def _create_object_fn(self, context: InitResourceContext) -> TResValue:
         return self.create_resource(context)
 
+    def get_context(self) -> InitResourceContext:
```
So one point of potential confusion here is that there will be a get_context() method in contexts where an OpExecutionContext is also available

e.g.

```
def an_asset(context: OpExecutionContext, a_resource: AResource):
   a_resource.get_context() != context
```

Maybe we should call it `get_resource_context` or `get_context_for_resource`

schrockn replied:
There are definitely problems with those names too

jamiedemaria replied:
are there cases when you would want/need to use the resource's context in an asset/op/etc? If not, then it could just be a private `_get_context` (or `_get_resource_context` etc) method right?

schrockn replied:
I don't think that users will actually call `get_context` in the body of the op. It's more about the conceptual confusion, which I still think would exist with a private `_get_context` method. Meaning that if I type that code, which context am I getting? Is it the same as the one passed to the op? It's already confusion and this risks making it more confusing.

jamiedemaria replied:
cool, that all makes sense to me!

schrockn replied:
`get_resource_init_context` potentially confusing as it is no longer access in an initializing context. It is accessed during execution

schrockn reviewed (changes requested):
Back to you to discuss `get_context` name

benpankow commented:
Updated to simply `get_resource_context`. Added a `with_resource_context` utility for testing purposes to let a user bind a context to a direct-instantiation resource as well.

schrockn commented on python_modules/dagster/dagster/_config/pythonic_config/__init__.py:
```diff
@@ -583,6 +584,7 @@ def __init__(self, **data: Any):
             config_schema=_curry_config_schema(schema, resolved_config_dict),
             schema=schema,
             nested_resources={k: v for k, v in resource_pointers.items()},
+            context=None,
```
`resource_context` to be consistent with this?

schrockn commented on python_modules/dagster/dagster/_config/pythonic_config/__init__.py:
```diff
@@ -697,15 +703,31 @@ def _resolve_and_update_nested_resources(
         return self._with_updated_values(to_update)
 
+    def with_resource_context(
+        self, context: InitResourceContext
```
resource_context

schrockn commented on python_modules/dagster/dagster/_config/pythonic_config/__init__.py:
```diff
@@ -697,15 +703,31 @@ def _resolve_and_update_nested_resources(
     def initialize_and_run(self, context: InitResourceContext) -> TResValue:
         with_nested_resources = self._resolve_and_update_nested_resources(context)
-        with_env_vars = with_nested_resources._resolve_and_update_env_vars()  # noqa: SLF001
+        with_context = with_nested_resources.with_resource_context(context)
+        with_env_vars = with_context._resolve_and_update_env_vars()  # noqa: SLF001
```
seems like this is a spot where we can do a one liner and it would be better

schrockn reviewed (changes requested):
being anal here, but let's change context to resource_context everywhere where appropriate, even in internals

schrockn commented on python_modules/dagster/dagster/_config/pythonic_config/__init__.py:
```diff
@@ -737,15 +743,35 @@ def _resolve_and_update_nested_resources(
+        self, resource_context: InitResourceContext
+    ) -> "ConfigurableResourceFactory[TResValue]":
+        """Returns a new instance of the resource with the given resource init context bound."""
+        copy = self._with_updated_values({})
```
this looks odd enough to be comment worthy

schrockn commented on python_modules/dagster/dagster/_config/pythonic_config/__init__.py:
```diff
@@ -737,15 +743,35 @@ def _resolve_and_update_nested_resources(
+        )
+        return copy
+
     def initialize_and_run(self, context: InitResourceContext) -> TResValue:
```
not your diff, but should we make this private since we don't want users to call this directly?
